### ISSUE TITLE
Why does DisablePrivateReflection attribute not work on .net core 2.1, but works perfectly on .net core 2.0?

### ISSUE THREAD

**Issue by monokot** [dotnet/coreclr#21985]
Hello!

Why does `DisablePrivateReflection` attribute not work on .net core 2.1, but works perfectly on .net core 2.0?

I have a class with private fields. The class is in assembly marked with a `DisablePrivateReflection` attribute.

The following test will be passed successfully on .net core 2.0

```csharp
var field = typeof(MonLicenseManager).GetField("_licenses", BindingFlags.NonPublic | 
BindingFlags.Static);

Assert.IsNotNull(field);

try
{
    var value = field.GetValue(null);
    Assert.Fail("FieldAccessException must be thrown");
}
catch (FieldAccessException)
{

}
```

but on .net core 2.1, `FieldAccessException` will never be thrown

Thanks

**Comment by ajcvickers** [comment]
@jkotas Two questions:
* Is this actually useful in any scenarios?
* Will attempts to create a compiled expression that accesses the field also fail?

**Comment by jkotas** [comment]
> Is this actually useful in any scenarios?

This was meant to be used by library developers who do not wish to allow external code to hack the library internals. The alternative is obfuscation that comes with its own set of issues.

> Will attempts to create a compiled expression that accesses the field also fail?

Was this case allowed in the .NET Framework implementation of this attribute?

I think we would try to match the .NET Framework behavior by default. However, I am not sure whether we would be able to match it exactly since this check was attached to code-access-security (CAS) that does exist in .NET Core.

**Comment by jkotas** [comment]
Right, this attribute was a speed bump - in the same way as obfuscation is a speed bump.

**Comment by ajcvickers** [comment]
@jkotas Okay, thanks for the info.

**Comment by terrajobst** [comment]
@jkotas 

Assuming we honor the attribute, how much of a performance penalty do we pay for success cases (i.e. when the attribute isn't applied)? In other words, is this an item where the majority has to pay for a feature only a few people will ever use?

**Comment by jkotas** [comment]
If we bring it back, it should have zero impact on the regular case (ie 100% pay-for-play).

It is why I have said that we may not be able to match the .NET Framework behavior exactly. We would not want to reintroduce the complex and expensive CAS checks from .NET Framework to match the behavior exactly.

**Comment by abatishchev** [comment]
Long story short: please don't bring it back. Imo, it should've never been existed in the first place.

**Comment by danmosemsft** [comment]
@jkotas @ajcvickers I'm moving this to Future as it does not seem essential for 3.0. Feel free to move back if I'm wrong.

**Comment by ajcvickers** [comment]
@danmosemsft Not in favor of it at all, so happy for you to punt it. :-)
